### PR TITLE
Make PostVisitBinaryOpNodeAddition more functional

### PR DESCRIPTION
**Pull request by bryancroteau-MSFT**
A big one, this changes PostVisitBinaryOpNodeAddition to be a mapping of inputs to outputs. This change also makes BinderCheckTypeResult a class instead of a readonly struct. This allows the Coercions property to default to an empty list instead of defaulting to null, removing the need to check for null when consuming the coercions list.

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs [hunk 1]
```diff
@@ -9,12 +9,34 @@ namespace Microsoft.PowerFx.Core.Binding
...
 
         // The binder has an imperative function called SetCoercedType,
         // which may be called multiple times in a given type checking function
         // This collection represents repeated calls to that function
-        public IReadOnlyCollection<BinderCoercionResult> Coercions { get; init; }
+        public IReadOnlyCollection<BinderCoercionResult> Coercions
+        {
+            get
+            {
+                if (_coercions == null)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Binding/BinderCheckTypeResult.cs:23]: Extremely minor nit, feel free to ignore. 
`return _coercions ?? new List<BinderCoercionResult>();`

**Review by lesaltzm — DISMISSED** [review]

**Review by lesaltzm — APPROVED** [review]
🚢

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:x: 1 Binary Compatibility issues for **Microsoft.PowerFx.Interpreter** compared against [version 0.2.2-preview.20220830-319453](https://www.nuget.org/packages/Microsoft.PowerFx.Interpreter/0.2.2-preview.20220830-319453).
<details>

```
MembersMustExist : Member 'Microsoft.PowerFx.PowerFxConfigExtensions.EnableCollectFunction(Microsoft.PowerFx.SymbolTable)' does not exist in the implementation but it does exist in the contract.
```

</details>
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**